module.c

```diff
@@ -59,6 +72,36 @@ static void teardown_time_limit(ContextData *context) {
 	}
 }
 
+// This method is always called in a context before running JS code in QuickJS. It sets up time
+// limites, releases the GIL etc.
+static void prepare_call_js(ContextData *context) {
+	// We release the GIL in order to speed things up for certain use cases.
+	assert(!context->thread_state);
+	context->thread_state = PyEval_SaveThread();
+	setup_time_limit(context, &context->interrupt_data);
```

Comment:
Time limits seem to be broken with the change to allow re-entrancy. I'm not sure whether the right behavior is to attribute time to the innermost script, to all scripts, or to the innermost script only while JS is running, or to forbid reentrancy when there is a time limit set. But the current behavior is that a nested JS call will call setup_time_limit and reset the timer, and then teardown_time_limit when it finishes which removes the time limit for the outer JS call.

Reply:
My 2 cents worth is that my expectation as a user of the API is that the time limit would apply to the originating script and that time would incremental for all time spent in JS until the termination of the originating script. Calls back to Python conceptually might be considered 'blocking' in JS space, and not allocate time to the JS execution time. Essentially, and again only conceptually, the time limit is a way of throttling the CPU cycles consumed in JS space. Of course, I might be wrong here, and Python time might be equally sensitive - a reasonable rationale for executing any untrusted code. The challenge is that it is probably not possible to differentiate the time spent in blocking vs non-blocking code in Python space.

Reply:
I'll disallow time limits and calling into Python for now.